### DISCUSSION TITLE
Can a master server be rebuilt?

### DISCUSSION THREAD

**Discussion by Stitch10925** [seaweedfs/seaweedfs#2938] [Q&A] [answered]
When a master server crashes, can it be rebuilt?

Let's say, for example, that the hard disk dies, or the master database gets corrupted, is there a way to rebuild this information using the volume servers? The data is obviously still there, the master just doesn't know how to access it.

**Comment by chrislusf** [comment, marked as answer]
The master only has soft states. Restart should be fine.

**Reply by Stitch10925** [reply]
Thank you for the fast reply!

I am still not able to access any data. Let me explain what I did, just to prevent us from misunderstanding each other:

(NOTE: This is all a test-setup, so I don't care if I cannot get the data back at this point. I am just testing out various failure scenario's).

On the master server I have a filer running on port 8888, a FUSE mount and a Samba server to share the FUSE mount for Windows systems. On the master server in the root there is a "filerldb2" file. This file I deleted to simulate a "corrupt" disk. I then restarted the server.

The master came online, filer came online and the filerldb2 file was recreated, FUSE mount and Samba share came online. However, when browsing the filer or samba share they are both empty. On the volumes everything is still the same, so the data is still there.

Now I am trying to get the filer and mount to show the data again. I assume I have to recreate or provision the filerldb2 file again somehow?

**Reply by chrislusf** [reply]
the filer data store needs to be backed up. It can not be rebuilt from the volume servers, which only has file content but not file path.

To support file path renaming, the volume servers would not store the file path information.

**Reply by Stitch10925** [reply]
I assume the filer gets it's data from the master, so can the master be used to rebuild te filer store? Because you can have a master and add a filer at a later point, so I would assume the master can give the filer the needed information.

**Reply by Stitch10925** [reply]
I did some digging into the shell and found the volume.fsck command:

> lock; volume.fsck; unlock
volume:5        entries:726     orphan:726      100.00% 1471407367B
volume:8        entries:829     orphan:829      100.00% 1723587486B
volume:2        entries:763     orphan:763      100.00% 1544032642B
volume:7        entries:814     orphan:814      100.00% 1692118725B
volume:6        entries:707     orphan:707      100.00% 1440431563B
volume:3        entries:752     orphan:752      100.00% 1508716406B
volume:1        entries:713     orphan:713      100.00% 1444194140B
volume:4        entries:788     orphan:788      100.00% 1592430170B

So it seems that all files are "orphaned", how do I "un-orphan" them?

**Reply by LuckyFrost** [reply]
>I assume the filer gets it's data from the master

This is an erroneous assumption. Master does not store path information

>Because you can have a master and add a filer at a later point, so I would assume the master can give the filer the needed information.

You can't, because without a filer, you have nowhere to store information about the paths, unless, of course, you yourself did not store them anywhere.

>So it seems that all files are "orphaned", how do I "un-orphan" them?

If you do not have information from the filer, then they cannot be returned.
At best you can get unstructured files (no directories, "all in one folder")

**Reply by Stitch10925** [reply]
So directory structures are purely a Filer and FUSE feature and are stored in their respective datastores?

If that is the case it should, in my opinion, be mentioned more specifically, at least in the "Production Setup" page that a Filer is potentially a single point of failure if there is only one of them or if the Async Filer Metadata Backup is not enabled. This was not clear to me at all.

Just out of curiosity, how would you get the files back even if the directory structure is lost? I assume a business would rather have their data back in an unorganized fashion than not at all.

It might be a nice feature to have a shell command to "un-orphan" the files in a volume by copying them or exporting them to somewhere else, depending on how difficult it is to get the files back manually.

**Reply by LuckyFrost** [reply]
>So directory structures are purely a Filer and FUSE feature and are stored in their respective datastores?
If that is the case it should, in my opinion, be mentioned more specifically, at least in the "Production Setup" 

https://github.com/chrislusf/seaweedfs/wiki/Components
This is the first item in the wiki.

"Production Setup" page also says that for production you need to have more than one filer if you need this service of course.

**Reply by Stitch10925** [reply]
I don't know why I missed that, thank you so much!

Any info on how to retrieve the orphaned files (without folder structuur)?
